src/toxcore/dht_new/server.rs

```diff
@@ -0,0 +1,630 @@
+- ping timeout lists ([`TimeoutQueue`])
+
+# Adding node to Close List
+
+Before a [`PackedNode`] is added to the Close List, it needs to be
+checked whether:
+
+- it can be added to [`Kbucket`] \(using [`Kbucket::can_add()`])
+- [`PackedNode`] is actually online
+
+Once the first check passes node is added to the temporary list, and
+a [`NodesRequest`] request is sent to it in order to check whether it's
+online. If the node responds correctly within [`PING_TIMEOUT`], it's
+removed from temporary list and added to the Close List.
+
+[`NodesRequest`]: ../dht/struct.NodesRequest.html
+[`Kbucket`]: ../dht/struct.Kbucket.html
+[`Kbucket::can_add()`]: ../dht/struct.Kbucket.html#method.can_add
+[`PackedNode`]: ../dht/struct.PackedNode.html
+[`PING_TIMEOUT`]: ../timeout/constant.PING_TIMEOUT.html
+[`TimeoutQueue`]: ../timeout/struct.TimeoutQueue.html
+*/
+#[derive(Clone)]
+pub struct Server {
+    /// secret key
+    pub sk: SecretKey,
+    /// public key
+    pub pk: PublicKey,
+    /// Close List (contains nodes close to own DHT PK)
+    pub kbucket: Kbucket,
+    /// tx split of channel to send packet to this peer via udp socket
+    pub tx: Tx,
+}
+
+impl Server {
+    /**
+    Create new `Server` instance.
+
+    Note: a new instance generates new DHT public and secret keys.
+
+    DHT `PublicKey` and `SecretKey` are supposed to be ephemeral.
+    */
+    pub fn new(tx: Tx) -> Result<Server, Error> {
+        if !crypto_init() {
+            return Err(Error::new(ErrorKind::Other,
+                       "Crypto initialization failed."));
+        }
+
+        let (pk, sk) = gen_keypair();
+        let kbucket = Kbucket::new(KBUCKET_BUCKETS, &pk);
+
+        debug!("Created new Server instance");
+        Ok(Server {
+            sk: sk,
+            pk: pk,
+            kbucket: kbucket,
+            tx: tx,
+        })
+    }
+
+    /// get pk of DhtPacket
+    pub fn get_pk(packet: &DhtPacket) -> PublicKey {
+        match packet.clone() {
+            DhtPacket::PingRequest(ping_request) => ping_request.pk,
+            DhtPacket::PingResponse(ping_response) => ping_response.pk,
+            DhtPacket::NodesRequest(nodes_request) => nodes_request.pk,
+            DhtPacket::NodesResponse(nodes_response) => nodes_response.pk,
+            DhtPacket::DhtRequest(dht_request) => dht_request.spk,
+            _ => panic!("Received packet is not a DhtPacket"),
+        }
+    }
+    /// create new client
+    pub fn create_client(&mut self, addr: &SocketAddr, pk: PublicKey) -> Client {
+        let precomputed_key = encrypt_precompute(&pk, &self.sk);
+        Client::new(precomputed_key, pk, addr.clone())
+    }
+    /**
+    Function to handle incoming packets. If there is a response packet,
+    send back it to the peer.
+    */
+    pub fn handle_packet(&mut self, packet: DhtUdpPacket) -> IoFuture<()>
+    {
+        let pk = Server::get_pk(&packet.1.clone());
+        let client = self.create_client(&packet.0, pk);
```

Comment:
Have a look at https://github.com/tox-rs/tox/pull/33 . You logic is incorrect for BootstrapInfo packet